pkg/config/crd/crdwatcher.go

```diff
@@ -0,0 +1,273 @@
+		}
+	}
+	w.mu.Unlock()
+	return keys, store.IndexNotSupported, nil
+}
+
+// Delete implements a KeyValueStore method.
+func (w *watcher) Delete(key string) error {
+	return errors.New("not implemented")
+}
+
+func unstructuredToResource(uns *unstructured.Unstructured) (string, *resource) {
+	md := uns.Object["metadata"].(map[string]interface{})
+	name := md["name"].(string)
+	version := md["resourceVersion"].(string)
+	return name, &resource{version, uns.Object["spec"]}
+}
+
+func (w *watcher) appendAndEncode(data resourceMap, field, key string) {
+	if len(data) == 0 {
+		return
+	}
+	res := make([]interface{}, 0, len(data))
+	for _, d := range data {
+		res = append(res, d.spec)
+	}
+	enc, err := yaml.Marshal(map[string]interface{}{field: res})
+	if err != nil {
+		glog.Errorf("Failed to marshal: %v", err)
+		return
+	}
+	w.data[key] = string(enc)
+}
+
+func (w *watcher) encodeData(data resourceMap, field string) {
+	for name, d := range data {
+		key := "/scopes/global/subjects/" + name + "/" + field
+		enc, err := yaml.Marshal(map[string]interface{}{field: d.spec})
+		if err != nil {
+			glog.Errorf("Failed to marshal: %v", err)
+			return
+		}
+		w.data[key] = string(enc)
+	}
+}
+
+func (w *watcher) buildDataAndNotify() {
+	w.mu.Lock()
+	defer w.mu.Unlock()
+	w.appendAndEncode(w.resources["Adapter"], "adapters", "/scopes/global/adapters")
+	w.appendAndEncode(w.resources["Handler"], "handlers", "/scopes/global/handlers")
+	if desc, ok := w.resources["Descriptor"]["global"]; ok {
+		enc, err := yaml.Marshal(desc.spec)
+		if err != nil {
+			return
+		}
+		w.data["/scopes/global/descriptors"] = string(enc)
+	}
+	w.encodeData(w.resources["Constructor"], "constructors")
+	w.encodeData(w.resources["Rule"], "rules")
+	w.encodeData(w.resources["Action"], "action_rules")
+	if w.listener != nil {
+		w.listener.NotifyStoreChanged(store.IndexNotSupported)
+	}
+}
+
+func (w *watcher) runWatch(wi watch.Interface, target resourceMap) {
+	defer wi.Stop()
+	for ev := range wi.ResultChan() {
+		if ev.Type == watch.Error {
+			glog.Errorf("watch error: %+v", ev.Object)
+			continue
+		}
+		name, d := unstructuredToResource(ev.Object.(*unstructured.Unstructured))
+		if ev.Type == watch.Deleted {
+			delete(target, name)
+		} else if prev, found := target[name]; !found || prev.version != d.version {
+			target[name] = d
+		} else {
+			continue
+		}
+		w.buildDataAndNotify()
+	}
+}
+
+func (w *watcher) init(config *rest.Config) error {
+	discovery, err := discovery.NewDiscoveryClientForConfig(config)
+	if err != nil {
+		return err
+	}
+	config.APIPath = "/apis"
+	config.GroupVersion = &schema.GroupVersion{Group: "config.istio.io", Version: "v1alpha2"}
+	dynamic, err := dynamic.NewClient(config)
+	if err != nil {
+		return err
+	}
+
+	list, err := discovery.ServerResourcesForGroupVersion("config.istio.io/v1alpha2")
+	if err != nil {
+		return err
+	}
+	for _, k := range []string{"Adapter", "Descriptor", "Constructor", "Handler", "Rule", "Action"} {
```

Comment:
Promote this list up to a var at the top of the file - maybe also declare vars for each of the strings since they're used in at least one other place as well (`buildDataAndNotify`).